plugins/teststeps/teststeps.go

```diff
@@ -27,54 +29,94 @@ type PerTargetFunc func(cancel, pause <-chan struct{}, target *target.Target) er
 // each target. The implementation of the per-target function is responsible for
 // handling internal cancellation and pausing.
 func ForEachTarget(pluginName string, cancel, pause <-chan struct{}, ch test.TestStepChannels, f PerTargetFunc) error {
-	for {
-		select {
-		case target := <-ch.In:
-			if target == nil {
-				// no more targets incoming
-				return nil
-			}
-			log.Debugf("%s: ForEachTarget: received target %s", pluginName, target)
-			errCh := make(chan error)
-			go func() {
-				log.Debugf("%s: ForEachTarget: calling function on target %s", pluginName, target)
-				errCh <- f(cancel, pause, target)
-			}()
+	type tgtErr struct {
+		target *target.Target
+		err    error
+	}
+	tgtErrCh := make(chan tgtErr)
+	tgtInFlight := int32(0)
+	noMoreTargetsCh := make(chan struct{})
 
+	go func(tgtInFlight *int32) {
+		defer func() {
+			log.Debugf("%s: ForEachTarget: exiting incoming loop, targets inflight %d", pluginName, tgtInFlight)
+		}()
+		for {
 			select {
-			case err := <-errCh:
-				if err != nil {
+			case tgt := <-ch.In:
+				log.Debugf("%s: ForEachTarget: received target %s", pluginName, tgt)
+				if tgt == nil {
+					log.Debugf("%s: ForEachTarget: all targets have been received", pluginName)
+					noMoreTargetsCh <- struct{}{}
+					return
+				}
+
+				go func() {
+					tgtErrCh <- tgtErr{target: tgt, err: f(cancel, pause, tgt)}
+				}()
```

Comment:
Ok I see. I think in the solution we were discussed in https://play.golang.org/p/Kvki6Z2RNHm, we were serializing read-write, i.e.

```
readValue <- chR
chW <- readValue
```

But I think this would work, just spawn a  background goroutine for each incoming target. Yeah, not sure why I haven't gone with this solution before, it seems so trivial now (I have everything in place, like workground, goroutine, etc. except that I serialize the goroutine...) :( 
I will give it a try, thank you.

Reply:
> I will give it a try, thank you.

I meant only to humbly ask to try to find a simpler solution for future code, I haven't meant to spent time on redoing what is already achieved :)